sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/StreamingProducer.cs

```diff
@@ -0,0 +1,520 @@
+            _partitionPublishers.Clear();
+            StartPartitionPublishersInternal(_options.MaximumQueuedEventCount);
+        }
+
+        /// <summary>
+        /// Performs the tasks needed to clean up all the resources used by the <see cref="StreamingProducer"/>, including ensuring that
+        /// the streaming producer itself has been closed.
+        /// </summary>
+        /// <remarks>
+        /// Calling this method will also <see cref="Clear(CancellationToken)"/> the queue, which will clear any events that are still pending,
+        /// and stop any active sending.
+        /// </remarks>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken" /> instance to signal the request to cancel the operation.</param>
+        public virtual async ValueTask CloseAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfClosed();
+
+            cancellationToken.ThrowIfCancellationRequested();
+            await FlushAsync(cancellationToken).ConfigureAwait(false);
+            await _producer.CloseAsync(cancellationToken).ConfigureAwait(false);
+            _isClosed = true;
+        }
+
+        /// <summary>
+        /// This method cancels any active sending and clears all events that are currently waiting to be sent in the queue.
+        /// </summary>
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken" /> instance to signal the request to cancel the operation.</param>
+        public virtual Task Clear(CancellationToken cancellationToken = default)
+        {
+            ThrowIfClosed();
+
+            foreach (var t in _cancellationTokens)
+            {
+                t.Cancel();
+            }
+
+            return Task.WhenAll(_partitionPublishers.Values);
+        }
+
+        /// <summary>
+        /// Performs the tasks needed to clean up all the resources used by the <see cref="StreamingProducer"/>, including ensuring that
+        /// the streaming producer itself has been closed.
+        /// </summary>
+        /// <remarks>
+        /// Calling this method will also clear the queue, which will drop any events that are still pending.
+        /// </remarks>
+        public ValueTask DisposeAsync()
+        {
+            ThrowIfClosed();
+            GC.SuppressFinalize(this);
+            return CloseAsync();
+        }
+
+        /// <summary>
+        /// This method is called upon the successful send of a batch of events. It calls the <see cref="SendSucceeded"/> event handler.
+        /// </summary>
+        /// <param name="events">An <see cref="IEnumerable{EventData}"/> instance holding all of the <see cref="EventData"/> instances within the
+        /// batch that was successfully sent.</param>
+        /// <param name="partitionId">The partition that the batch of events was sent to.</param>
+        protected virtual void OnSendSucceededAsync(IEnumerable<EventData> events, string partitionId)
```

Comment:
```suggestion
        protected virtual TaskOnSendSucceededAsync(IEnumerable<EventData> events, string partitionId)
```

We never want an async method to return `void`... that makes it impossible to await.